---
title: "Which samples match their expected parents?"
author: "Tom Ellis"
date: "2024-08-06"
output: html_document
---

This notebook scrapes the output of `snpmatch cross` comparing 429 F8s to 1163 candidate parents, to determine whether F8 individuals match the genotypes of who we expect the parents to be.

Starting material:

* `01_data/02_F8_unaligned_bams/sequencing_plates_original.csv` lists each plant as it was sent to be sequenced, giving plate and position, expected genotype, and sequencing filename, among other things.
* SNPmatch results are stored on scratch (will probably have gone by the time you read this). There is one directory per F8 line, named after the plate and well position (e.g. 2_A3 is the plant in plate 2, position A3, as it was sent to be sequenced).
* Each directory contains the following files, created by SNPmatch:
+ `2_A3.matches.json`: JSON file listing most likely parents, and all compatible parents, although the numbers are not completely clear.
+ `2_A3.scores.txt.matches.json`
+ `2_A3.vcf.snpmatch.stats.json`
+ `2_A3.scores.txt`: described in the [SNPmatch readme](https://github.com/Gregor-Mendel-Institute/snpmatch?tab=readme-ov-file#output-files-for-inbred)
+ `2_A3.vcf.snpmatch.npz`
+ `2_A3.windowscore.txt`: described in the [SNPmatch readme](https://github.com/Gregor-Mendel-Institute/snpmatch?tab=readme-ov-file#output-files-for-cross)

Plan:
* How many F8s have the correct parents as the top candidates?
* How many F8s have the correct parents within the list of compatible parents? Need to set a limit on this, and validate some visually.
* For the remainder, are the expected parents in Rahul's list of mixed up parents?
* Are there sets of orphaned genotypes who share the same missing parent? Could these have been mixed with someone else?


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

```{r dependencies}
library(tidyverse)
library(ggpubr)
library(jsonlite)
```

```{r most-likely-parents}
plate_info <- read_csv(
  "01_data/02_F8_unaligned_bams/sequencing_plates_original.csv",
  show_col_types = FALSE
) %>% 
  mutate(
    sampleID = paste0(plate, "_", row, col)
  ) %>% 
  separate(name, into = c("genotype", NA, NA), sep = " ", remove = FALSE) %>% 
  separate(genotype, into=c("exp_parent1", "exp_parent2"), sep = "x") %>% 
  select(sampleID, name, exp_parent1, exp_parent2)

# Find the JSON files ending in .matches.json, excluding those ending with scores.txt.matches.json
matches.json_files <- Sys.glob(
  "/scratch-cbe/users/thomas.ellis/crosses/03_validate_genotypes/03_run_SNPmatch/[0-5]_[A-H]*[!scores.txt].matches.json"
)

# Dataframe with only the sample IDs in the JSON files
sampleID_from_JSON <- tibble(
  sampleID = str_extract(
    basename(matches.json_files), pattern = "^._[A-H][0-9]{1,2}"
  )
)

# For each JSON file, pull out the most likely pair of parents
most_likely_parents <- vector('list', length(matches.json_files))
for(i in 1:length(matches.json_files)){
  matches <- fromJSON(matches.json_files[i])
  
  most_likely_parents[[i]] <- tibble(
    sampleID = str_extract(
      basename(matches.json_files[i]), pattern = "^._[A-H][0-9]{1,2}"
    ), 
    obs_parent1 = as.data.frame(matches$matches)$V1[1],
    obs_parent2 = as.data.frame(matches$matches)$V1[2]
  )
}

# Join expected parents to most likely parents, and check if they match
most_likely_parents <- most_likely_parents %>% 
  do.call(what='rbind') %>% 
  right_join(plate_info, by= 'sampleID') %>% 
  mutate(
    match_parent1 = ifelse(
      (exp_parent1 == obs_parent1) | (exp_parent1 == obs_parent2), TRUE, FALSE
    ),
    match_parent2 = ifelse(
      (exp_parent2 == obs_parent1) | (exp_parent2 == obs_parent2), TRUE, FALSE
    )
  )


n_matches_both_parents <- most_likely_parents %>%
  filter(match_parent1 & match_parent2) %>% nrow()

n_matches_one_parents <- most_likely_parents %>% 
  filter(xor(match_parent1, match_parent2)) %>%  nrow()

```

```{r}
na_parents <- most_likely_parents %>% 
  filter(is.na(obs_parent1)) %>% 
  pull(sampleID)

x <- c('2_G3','2_A3','2_F7','2_E9','2_B7','2_A11','2_G8','2_B10','2_F9','2_A9','2_G1','2_E4','2_C9','2_E3','2_H6','2_D12','2_C1','2_D4','2_A6','2_F12','2_G5','2_H4','2_G11','2_H9','2_B8','2_D2','2_B4','2_F10','2_C12','2_A1','2_C10','2_B12','2_G2','2_A5','2_C3','2_A2','2_H7','2_E12','2_H10','2_F5','2_D1','2_C2','2_C7','2_F3','2_B6','2_H2','2_C5','2_F1','2_B9','2_A12','2_F11','2_C8','2_E10','2_C6','2_G12','2_G9','2_B3','2_C11','2_D10','2_H3','2_D11','2_E5','2_A7','2_F2','2_A10','2_H8','2_F8','2_E1','2_F6','2_E2','2_H11','2_B5','2_D5','2_E6','2_G6','2_B1','2_G7','2_G10','2_D3','2_B2','2_D7','2_H5','2_E8','2_F4','2_H1','2_G4','2_H12','2_C4','2_D6','2_E7','2_A8','2_E11','2_D8','2_A4','2_B11','2_D9','4_E5','4_H6','4_E11','4_A9','4_F11','4_F6','4_A12','4_C7','4_C9','4_A4','4_G9','4_C12','4_H3','4_F4','4_C3','4_E4','4_C10','4_D2','4_F3','4_B5','4_C8','4_E1','4_H11','4_D3','4_F1','4_D5','4_A10','4_G2','4_F10','4_E12','4_H7','4_F12','4_B1','4_A2','4_F9','4_A8','4_H5','4_D7','4_H12','4_A6','4_E3','4_D10','4_H10','4_F8','4_B7','4_D11','4_C4','4_D9','4_H4','4_E8','4_A11','4_D4','4_E6','4_C2','4_C6','4_A3','4_B4','4_B11','4_E7','4_D8','4_B10','4_G4','4_G10','4_B6','4_E2','4_H1','4_B12','4_E10','4_H8','4_B8','4_G6','4_F2','4_D6','4_B3','4_F7','4_C11','4_H9','4_A7','4_D1','4_G3','4_G8','4_G1','4_A5','4_C1','4_H2','4_A1','4_D12','4_G5','4_F5','4_E9','4_B9','4_G11','4_C5','4_B2','4_G12','4_G7','3_G11','3_A2','3_F10','3_D10','3_F8','3_G6','3_D5','3_H4','3_B12','3_H11','3_E9','3_C11','3_B3','3_D12','3_C2','3_E6','3_D6','3_D7','3_G7','3_A8','3_F7','3_G10','3_G12','3_B8','3_G9','3_E7','3_A4','3_B9','3_H5','3_F12','3_A11','3_C3','3_H3','3_C12','3_B7','3_B1','3_E11','3_G1','3_B5','3_G3','3_D4','3_F4','3_C6','3_F2','3_B6','3_B10','3_D3','3_D11','3_G2','3_B11','3_H1','3_H2','3_C8','3_D1','3_A3','3_E5','3_C5','3_H6','3_D9','3_A9','3_D2','3_H12','3_E10','3_F11','3_F5','3_D8','3_C10','3_F1','3_A6','3_A10','3_G4','3_A5','3_E4','3_H10','3_A12','3_G8','3_C9','3_H7','3_E2','3_E1','3_E8','3_A1','3_H9','3_A7','3_G5','3_E12','3_H8','3_B2','3_C4','3_F9','3_E3','3_F3','3_C1','3_C7','3_F6','3_B4','1_D8','1_D11','1_F1','1_C5','1_D1','1_F6','1_C7','1_B10','1_E4','1_G8','1_B9','1_B6','1_H7','1_C8','1_B2','1_G10','1_B11','1_H12','1_D10','1_A3','1_E5','1_G6','1_D4','1_F2','1_C12','1_G4','1_E2','1_B7','1_H11','1_C6','1_C3','1_F3','1_D6','1_H5','1_B4','1_D5','1_B12','1_H2','1_E9','1_F7','1_H9','1_H8','1_H4','1_E6','1_H1','1_B5','1_A8','1_C11','1_E8','1_F11','1_C1','1_A1','1_A2','1_E7','1_A6','1_D9','1_A9','1_D7','1_A11','1_F8','1_H3','1_G2','1_D12','1_C4','1_D2','1_E10','1_A10','1_E11','1_B8','1_F10','1_B3','1_G12','1_F12','1_C9','1_H10','1_A5','1_E1','1_G9','1_G11','1_G7','1_A12','1_G1','1_F5','1_E12','1_A7','1_C10','1_B1','1_F4','1_E3','1_A4','1_F9','1_H6','1_D3','1_C2','1_G3','1_G5','5_A2','5_G6','5_D5','5_H4','5_B3','5_C2','5_E6','5_D6','5_A4','5_H5','5_C3','5_H3','5_B1','5_G1','5_B5','5_G3','5_D4','5_F4','5_C6','5_F2','5_D3','5_G2','5_H1','5_H2','5_D1','5_A3','5_E5','5_C5','5_H6','5_D2','5_F5','5_F1','5_G4','5_A5','5_E4','5_E2','5_E1','5_G5','5_B2','5_C4','5_E3','5_F3','5_C1','5_F6','5_B4')

match(na_parents, x)

```


If we pull only the two most likely parents from output of SNPmatch:

* `r n_matches_both_parents` samples match both expected parents.
* `r n_matches_one_parents` samples match one, but not both parents.

```{r}

parental_rank <- vector('list', length(matches.json_files))
for(i in 1:length(matches.json_files)){
  matches <- fromJSON(matches.json_files[i])
  
  sample_name = str_extract(
      basename(matches.json_files[i]), pattern = "^._[A-H][0-9]{1,2}"
    )
  this_row <- plate_info %>% 
    filter(sampleID == sample_name)
  parental_rank[[i]] <- tibble(
    sampleID= sample_name,
    rank_parent1 = match( this_row$exp_parent1, as.data.frame(matches$matches)$V1),
    rank_parent2 = match( this_row$exp_parent2, as.data.frame(matches$matches)$V1),
    compatible_candidates = paste(as.data.frame(matches$matches)$V1[1:20], collapse = " "),
    compatible_scores = paste(as.data.frame(matches$matches)$V2[1:20], collapse = " ")
  )
}

parental_rank <- parental_rank %>% 
  do.call(what='rbind') %>% 
  right_join(most_likely_parents, by= 'sampleID')

# 296 have both parents in the top 20
parental_rank %>% 
  filter(rank_parent1 <21 & rank_parent2 < 21)

write_csv(parental_rank, file="03_processing/03_validate_genotypes/output/snp_match_results.csv")

```

```{r}

parental_rank %>% 
  filter( xor(rank_parent1 >20, rank_parent2 >20) ) %>% 
  mutate(
    bad_parent = ifelse(rank_parent1 > 20, exp_parent1, exp_parent2)
    ) %>% 
  select( -match_parent1, -match_parent2) %>% 
  split(., .$bad_parent)
    
  

```



