---
title: "Flowering time GWAS across resampled datasets"
author: "Tom Ellis"
date: today
format:
  html:
    toc: true
    self-contained: true
---

```{r setup, echo=FALSE}
knitr::opts_chunk$set(echo = FALSE, cache = TRUE, cache.lazy=TRUE)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())

suppressPackageStartupMessages(library(tidyverse))
library(ggpubr)


```

# Sums of log10 p-values

```{r load-sum-pvals}
#| messages: false

source("02_library/manhattan_plot.R")


sum_pvals_noK <- read_csv(
  "05_results/19_ft12_resampled/output/ft12_resampled_sum_pvals_no_K.csv"
) %>% 
  filter(!is.na(sum_log_p)) %>% 
  mutate(
    SNP = paste0(chr, ":", ps),
    sum_log_p = sum_log_p
  )
sum_pvals_noK <- add_base_pair_positions(sum_pvals_noK)


sum_pvals_withK <- read_csv(
  "05_results/19_ft12_resampled/output/ft12_resampled_sum_pvals_with_K.csv"
) %>% 
  filter(!is.na(sum_log_p)) %>% 
  mutate(
    SNP = paste0(chr, ":", ps),
    sum_log_p = sum_log_p
  )
sum_pvals_withK <- add_base_pair_positions(sum_pvals_withK)


```

The following three plots show:

1. Sum of -log10 p values across datasets for models *including* a relatedness matrix
2. Sum of -log10 p values across datasets for models *without* a relatedness matrix
3. -log10 p values for the full F9 data set with >400 lines estimated *without* a relatedness matrix.

We see that:

1. Models with and without a relatedness matrix give very similar answers.
2. The peaks for resampled datasets match the peaks for the full F9 dataset
3. The peaks for resampled datasets match one or both observed cohorts of F9s (not actually shown here).

```{r sum-p-withK, fig.width=15, fig.height=8}
sum_pvals_withK %>% 
  filter(sum_log_p > 300) %>% 
  ggplot(aes(x=ps_cum, y=sum_log_p)) +
  geom_point( aes(color=as.factor(chr)), alpha=0.8, size=1.3) +
  scale_color_manual(values = rep(c("grey", "skyblue"), 22 )) +
  labs(
    x = "Chromosome",
    y = expression(paste("-log"[10],"p-value")),
    title = "With relatedness matrix"
  ) +
  theme_bw() +
  theme( 
    legend.position="none",
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank()
  )

```


```{r sum-p-noK, fig.width=15, fig.height=8}

sum_pvals_noK %>% 
  filter(sum_log_p > 500) %>% 
  ggplot(aes(x=ps_cum, y=sum_log_p)) +
  geom_point( aes(color=as.factor(chr)), alpha=0.8, size=1.3) +
  scale_color_manual(values = rep(c("grey", "skyblue"), 22 )) +
  labs(
    x = "Chromosome",
    y = expression(paste("-log"[10],"p-value")),
    title = "No relatedness matrix"
  ) +
  theme_bw() +
  theme( 
    legend.position="none",
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank()
  )

```


```{r load-all-F9s}
all_F9s_no_K <- read_delim(
  "05_results/16_gemma_ft12/output/no_K/flowering_time_blups_F9_combined.assoc.txt",
  delim = '\t',
  show_col_types = FALSE
) %>% 
  mutate(
    SNP = paste0(chr, ":", ps),
    log10p = -log10(p_lrt)
  )
all_F9s_no_K <- add_base_pair_positions(all_F9s_no_K)
```


```{r mh-allf9, fig.width=15, fig.height=8}
all_F9s_no_K %>% 
  filter(log10p > 5) %>% 
  ggplot(aes(x=ps_cum, y=log10p)) +
  geom_point( aes(color=as.factor(chr)), alpha=0.8, size=1.3) +
  scale_color_manual(values = rep(c("grey", "skyblue"), 22 )) +
  labs(
    x = "Chromosome",
    y = expression(paste("-log"[10],"p-value")),
    title = "All F9s at once"
  ) +
  theme_bw() +
  theme( 
    legend.position="none",
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank()
  )

```


# Effect sizes

```{r load-sum-betas}
#| messages: false

sum_betas_noK <- read_csv(
  "05_results/19_ft12_resampled/output/ft12_resampled_mean_betas_K_no_K.csv",
  col_types = 'iid'
  ) %>% 
  filter(!is.na(sum_beta)) %>% 
  mutate(
    SNP = paste0(chr, ":", ps),
    mean_beta = sum_beta / 200
  )
sum_betas_noK <- add_base_pair_positions(sum_betas_noK)


sum_betas_withK <- read_csv(
  "05_results/19_ft12_resampled/output/ft12_resampled_mean_betas_K_with_K.csv",
  col_types = 'iid'
) %>% 
  filter(!is.na(sum_beta)) %>% 
  mutate(
    SNP = paste0(chr, ":", ps),
    mean_beta = sum_beta / 200
  )
sum_betas_withK <- add_base_pair_positions(sum_betas_withK)


```


```{r mean-beta-noK, fig.width=15, fig.height=8}
sum_betas_noK %>% 
  # filter(sum_log_p > 300) %>%
  ggplot(aes(x=ps_cum, y=abs(mean_beta))) +
  geom_point( aes(color=as.factor(chr)), alpha=0.8, size=1.3) +
  scale_color_manual(values = rep(c("grey", "skyblue"), 22 )) +
  labs(
    x = "Chromosome",
    y = expression(paste("Mean absolute effect size")),
    title = "No relatedness matrix"
  ) +
  theme_bw() +
  theme( 
    legend.position="none",
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank()
  )

```

```{r mean-beta-withK, fig.width=15, fig.height=8}
sum_betas_withK %>% 
  # filter(sum_log_p > 300) %>%
  ggplot(aes(x=ps_cum, y=abs(mean_beta))) +
  geom_point( aes(color=as.factor(chr)), alpha=0.8, size=1.3) +
  scale_color_manual(values = rep(c("grey", "skyblue"), 22 )) +
  labs(
    x = "Chromosome",
    y = expression(paste("Mean absolute effect size")),
    title = "Including relatedness matrix"
  ) +
  theme_bw() +
  theme( 
    legend.position="none",
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank()
  )

```
